### PR TITLE
fix child rebuild under layoutbuild cause duplicate globalkey error

### PR DESCRIPTION
**Pull request by chunhtai**
## Description

The layoutbuilder delays its rebuild to rendering time. If a child under it is marked dirty, it will rebuild with outdated widget that will eventually updated and rebuild during layout.
This is generally ok except this will pose a problem for global key reservation.
This pr attempts to remove global key reservation that cause by dirty node rebuild prior to layoutbuilder rebuild

## Related Issues

https://github.com/flutter/flutter/issues/43780

## Tests

I added the following tests:

"GlobalKey - can reorder under LayoutBuilder while dirtying children"

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I signed the [CLA].
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -184,6 +184,13 @@ abstract class GlobalKey<T extends State<StatefulWidget>> extends Key {
           ),
         ]);
       }
+      // The parent might have multiple global key reservations. This can happen
+      // when a dirty node under LayoutBuilder is rebuilt before the LayoutBuilder
+      // updates its widget subtree in rendering layer. At the end, the result
+      // of dirty node rebuild will be overwritten by the LayoutBuilder rebuild.
+      // It should be safe to remove all the other reservations of GlobalKeys.
+      // See https://github.com/flutter/flutter/issues/43780
+      _debugReservations.removeWhere((_, Element e) => e == parent);
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/framework.dart]: nit: avoid using `_`, give the actual type and name of the argument.

File: packages/flutter/test/widgets/framework_test.dart [hunk 2]
```diff
@@ -539,6 +539,54 @@ void main() {
...
+        children.add(listItemToWidget(i, listItems[i]));
+      }
+      return Column(
+        children: children,
+      );
+    }
+
+    await tester.pumpWidget(
+      LayoutBuilder(
+        builder: (_,__) => buildContents(),
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart]: ditto

File: packages/flutter/test/widgets/framework_test.dart [hunk 3]
```diff
@@ -539,6 +539,54 @@ void main() {
...
+    // Marks one child dirty.
+    state.rebuild();
+    // Reorders the children.
+    final String temp = listItems[1];
+    listItems[1] = listItems[2];
+    listItems[2] = temp;
+    // This will mark the root Element to be dirty and pump a new frame.
+    await tester.pumpWidget(
+      LayoutBuilder(
+        builder: (_,__) => buildContents(),
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart]: ditto

File: packages/flutter/test/widgets/framework_test.dart [hunk 4]
```diff
@@ -539,6 +539,54 @@ void main() {
...
+    Widget listItemToWidget(int index, String listItem) {
+      return _Stateful(
+        child: Text(
+          listItem,
+          textDirection: TextDirection.ltr,
+          key: GlobalObjectKey(listItem),
+        ),
+      );
+    }
+    Widget buildContents() {
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart]: if you just gave this two arguments, we could avoid making closures below

File: packages/flutter/test/widgets/framework_test.dart [hunk 5]
```diff
@@ -539,6 +539,54 @@ void main() {
...
+    listItems[2] = temp;
+    // This will mark the root Element to be dirty and pump a new frame.
+    await tester.pumpWidget(
+      LayoutBuilder(
+        builder: (_,__) => buildContents(),
+      )
+    );
+    expect(tester.takeException(), null);
+  });
+
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart]: also create some tests that check that you get an exception in the various cases where you do end up with duplicate globalkeys when dealing with a layout builder. (e.g. when the order is changed to include the same item twice)

File: packages/flutter/test/widgets/framework_test.dart [hunk 6]
```diff
@@ -754,3 +802,20 @@ class DirtyElementWithCustomBuildOwner extends Element {
...
+class _Stateful extends StatefulWidget {
+  const _Stateful({Key key, this.child}) : super(key: key);
+  final Widget child;
+
+  @override
+  State<StatefulWidget> createState() => _StatefulState();
+}
+
+class _StatefulState extends State<_Stateful> {
+  void rebuild() => setState(() {});
```
Comment (Hixie) [inline packages/flutter/test/widgets/framework_test.dart]: always put a comment inside empty setState

**Comment by goderbauer** [comment]
I was playing around with the test case a little bit to understand how this fix is supposed to work and it seems like it still crashes with a slight variation of the test case:

```dart
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

void main() {
  group('KeyTest', () {
    final List<String> listItems = <String>['Item 1', 'Item 2'];

    Widget listItemToWidget(int index, String listItem) {
      return _Stateful(
        child: Text(
          listItem,
          textDirection: TextDirection.ltr,
          key: GlobalObjectKey(listItem),
        ),
      );
    }

    Widget buildContents() {
      final List<Widget> children = <Widget>[];
      for (int i = 0; i < listItems.length; i++) {
        children.add(listItemToWidget(i, listItems[i]));
      }
      return Column(
        children: children,
      );
    }

    testWidgets('allows reordering inside the middle of the widget (reorder up)', (WidgetTester tester) async {
      await tester.pumpWidget(
          LayoutBuilder(
            builder: (_,__) => buildContents(),
          )
      );
      final _StatefulState state = tester.firstState(find.byType(_Stateful).at(1));
      state.rebuild();
      // Reorder the items
      final String temp = listItems[0];
      listItems[0] = listItems[1];
      listItems[1] = temp;
      await tester.pumpWidget(
          LayoutBuilder(
            builder: (_,__) => buildContents(),
          )
      );
    });
  });
}

class _Stateful extends StatefulWidget {
  const _Stateful({Key key, this.child}) : super(key: key);
  final Text child;
  @override
  State<StatefulWidget> createState() => _StatefulState();
}

class _StatefulState extends State<_Stateful> {
  void rebuild() => setState(() {});
  @override
  void didUpdateWidget(_Stateful oldWidget){
    super.didUpdateWidget(oldWidget);
    print('old widget ${oldWidget.child.data} to new widget ${widget.child.data}');
  }
  @override
  Widget build(BuildContext context) {
    return widget.child;
  }
  @override
  String toString({DiagnosticLevel minLevel = DiagnosticLevel.debug}) {
    return 'State ${widget.child.data}';
  }
}
```

**Comment by chunhtai** [comment]
It looks like there is a different corner case i didn't think of

**Comment by goderbauer** [comment]
Looks like the order in which keys get claimed matter.

Let's say we have keys [k1, k2] and element [eA, eB] as children of the layout builder.

When you rebuild eA and then switch the list the following happens (this is the test case included in the PR):
eA claims k1 (rebuild of eA)
eA claims k2 [& unclaims k1], eB claims k1 (rebuild of the layout builder).
Everything is fine here with the unclaim fix.

When you rebuild eB and then switch the list the following happens:
eB claims k2 (rebuild of eB)
eA claims k2, eB claims k1 (rebuild of the layout builder)
This case still crashes because the newly introduced reservation removal code has nothing to remove.

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (krisgiesing) [inline packages/flutter/lib/src/widgets/framework.dart]: Suppose I have erroneous code where parent A has children with keys X and Y, and I write code that erroneously reuses key X for parent B. As I understand it, this code will remove X from the reservation table when Y is processed, thus preventing me from seeing an exception when I then put X under B.

Am I misunderstanding?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/framework.dart]: Yes, you are right. This is not going to work. I will probably have to come up a better solution

**Comment by chunhtai** [comment]
I think we are in a situation where layoutbuilder rebuild its children and try to reserve the globalkey.
It cannot tell whether the previously reservation to that globalkey is a result of a real globalkey duplication or an early rebuild with the old widget.

The reservation removal does not really work as well. it does not know if the other reservation is valid or a result of early rebuild.

We cannot drop the early rebuild either, because we cannot guarantee the layoubuilder will actually rebuild that dirty node. It might terminate early due widget tree does not change.

It seems like some kind of refactoring of globalkey reservation check is needed. I will keep investigating

**Comment by goderbauer** [comment]
Should we close this PR for now since we will have to chose a different approach?

**Pull request closed without merging**